### PR TITLE
Add an explicit fallback to net_http to support both faraday 1.x and 2.x implementations

### PR DESCRIPTION
**Pull request by cbeer**

### CODE REVIEW

File: rsolr.gemspec [hunk 1]
```diff
@@ -27,13 +27,14 @@ Gem::Specification.new do |s|
...
   s.require_paths = ["lib"]
-  
+
   s.required_ruby_version      = '>= 1.9.3'
-  
+
   s.requirements << 'Apache Solr'
 
   s.add_dependency 'builder', '>= 2.1.2'
   s.add_dependency 'faraday', '>= 0.9.0'
```
Comment (cjcolvar) [inline rsolr.gemspec:36]: FWIW it looks like the faraday dependency will never be able to <1 due to faraday-net_http's requirements.  Should this be bumped up to `>= 1.0` to not give false hopes?

File: lib/rsolr/client.rb [hunk 2]
```diff
@@ -319,7 +320,7 @@ def connection
         conn.request :retry, max: options[:retry_after_limit], interval: 0.05,
                              interval_randomness: 0.5, backoff_factor: 2,
                              exceptions: ['Faraday::Error', 'Timeout::Error'] if options[:retry_503]
-        conn.adapter options[:adapter] || Faraday.default_adapter
+        conn.adapter options[:adapter] || Faraday.default_adapter || :net_http
```
Comment (jcoyne) [inline lib/rsolr/client.rb:323]: Does Faraday.default_adapter return nil in 2.x? Why do they even have that method?

Comment (cbeer) [reply lib/rsolr/client.rb:323]: According to the upgrade guide, downstream consumers can set default_adapter.

File: rsolr.gemspec (hunk 1 above)
Comment (jrochkind) [reply rsolr.gemspec:36]: I don't think that's true -- `faraday-net_http` is only a dependency of faraday depending on what version of faraday you are using. If you are using faraday < 1, you don't have `faraday-net_http` at all. I think there's no reason for rsolr to stop letting users use a version of faraday they may be currently using without problems -- especially on a minor version release.

Comment (cjcolvar) [reply rsolr.gemspec:36]: @jrochkind `faraday-net_http` is being required here on the next line.  From what I can see in rubygems, `faraday-net_http` has only every supported `faraday` 1.0 and higher.  By adding a hard dependency on `faraday-net_http` here then rsolr would essentially be requiring faraday 1.0+.

**Review by jcoyne — APPROVED** [review]

**Comment by jrochkind** [comment]
Note this PR changed our dependency to require faraday 1.0 instead of 0.9.0.  Personally I would argue that should not be done in a minor release. 

But no other committers think this/are concerned about it? I can get overruled. Would have liked to sit on this a *bit* longer for discussion -- is there any reason we need to be merging things effecting backwards compat the same hour they are submitted, when they are receiving discussion?

File: rsolr.gemspec (hunk 1 above)
Comment (jrochkind) [reply rsolr.gemspec]: I think we could have done this differently to preserve backwards compat with all previously supported faraday versions -- we didn't need to explicitly depend on `faraday-net_http`, and didn't need to do the `require` if Faraday was less than 2, which we could have checked for. 

I would have appreciated y'all giving this more time for discussion when we are in the middle of a discussion -- what's the rush/urgency?

**Comment by jcoyne** [comment]
> is there any reason we need to be merging things effecting backwards compat the same hour they are submitted, when they are receiving discussion?

I didn't see this as effecting backward compatibility.  It looked to me like all points raised were addressed.

**Comment by jcoyne** [comment]
> Note this PR changed our dependency to require faraday 1.0 instead of 0.9.0. Personally I would argue that should not be done in a minor release.

I think this is a bug fix, because we previously said any version of faraday >= 0.9.0  would work, but if you use faraday 2, it breaks.

**Comment by jrochkind** [comment]
Ah, I understand the urgency now, `>= 9.0` meant builds would break. 

I still wish we had tried to find a way to not change the lower bound of faraday we support. But I'm also seeing faraday [isn't doing what I thought it was going to do here](https://github.com/lostisland/faraday/issues/1364), so the path I was thinking doesn't exactly work. 

I don't myself use faraday 0.9.0, this won't be a problem for me. Just in this thing that is *very* "legacy" and used outside our community in who knows what ways, I think we have a responsibility to be super careful with backwards compat. But hopefully it will be fine. 

I wonder if we should set an *upper* bound of `< 3` to avoid running into this problem again?  I think it's most responsible to always set major version upward bounds for dependencies using semver past 1.0 (which faraday now is), since it is expected the next major version will have backwards incompat of some kind, right?

**Comment by jcoyne** [comment]
> I wonder if we should set an upper bound of < 3 to avoid running into this problem again? I think it's most responsible to always set major version upward bounds for dependencies using semver past 1.0 (which faraday now is), since it is expected the next major version will have backwards incompat of some kind, right?

Yeah, that sounds like a good idea.

**Comment by cjcolvar** [comment]
I would +1 making this a major release since as far as I can tell most Samvera gems aren't even on faraday 1 yet.

**Comment by iMacTia** [comment]
Hey! I've just released Faraday 2.0.1 which brings back `net_http` as the default adapter.
It would be great if you could give it a try and let me know if that fixes your issue?

**Comment by jrochkind** [comment]
We may be able to do undo this becuase of the faraday correction @iMacTia mentions?

> I've just released Faraday 2.0.1 which brings back faraday-net_http as a dependency and sets it as the default adapter 👍

https://github.com/lostisland/faraday/discussions/1358#discussioncomment-1915353

We're still going to have a problem with the `retry_503` option in #230 though....

**Pull request merged**